File: 00-code(源代码)/src/com/hi/dhl/algorithms/_270/java/Solution.java
package com.hi.dhl.algorithms._270.java;

import java.util.*;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/1
 *     desc  :
 * </pre>
 */
class Solution {
    public int numSquares(int n) {
        List<Integer> square = new ArrayList<>();
        for (int i = 1; i < n; i++) {
            square.add(i * i);
        }

        int step = 0;
        Set<Integer> set = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        queue.add(n);
        while (queue.size() > 0) {
            step = step + 1;
            int count = queue.size();
            for (int i = 0; i < count; i++) {
                int item = queue.poll();
                for (Integer subItem : square) {
                    if (item == subItem) {
                        return step;
                    } else if (item < subItem) {
                        break;
                    } else {
                        int sub = item - subItem;
                        if (set.add(sub)) {
                            queue.offer(sub);
                        }
                    }
                }
            }
        }
        return step;
    }

    public static void main(String... args) {
        Solution solution = new Solution();
        System.out.println(solution.numSquares(1));
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/_270/kotlin/Solution.kt
package com.hi.dhl.algorithms._270.kotlin

import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/1
 *     desc  :
 * </pre>
 */

class Solution {
    fun numSquares(n: Int): Int {
        val square = mutableListOf<Int>()
        for (i in 1..n) {
            square.add(i * i)
        }

        val set = hashSetOf<Int>()
        val queue = LinkedList<Int>()
        queue.offer(n)
        var step = 0
        while (queue.size > 0) {
            step = step + 1
            val count = queue.size
            for (i in 0 until count) {
                val item = queue.poll()
                loop@ for (j in square) {
                    when {
                        item == j -> return step
                        item < j -> break@loop
                        else -> {
                            val sub = item - j;
                            if (set.add(sub)) {
                                queue.offer(sub)
                            }
                        }
                    }
                }
            }
        }
        return step
    }
}